www/lib/searchDescriptor.ts

```diff
@@ -0,0 +1,91 @@
+import { Class, Func, Module, Project, Variable } from "./docs";
+import {
+  mod as generateModuleUrl,
+  project as generateProjectUrl,
+  symbol as generateSymbolUrl,
+} from "./url";
+
+export type SearchDescriptor = {
+  name: string;
+  // TODO: finish this list
+  type?: "project" | "module" | "function" | string;
+  url: string;
+};
+
+const generateSymbolDescriptors = (
+  descriptors: Array<SearchDescriptor>,
+  project: Project,
+  module: Module,
+  symbols: Array<Class | Func | Variable>,
+  symbolType: string
+) => {
+  symbols.forEach((symbol) => {
+    descriptors.push({
+      name: symbol.name,
+      type: symbolType,
+      url: generateSymbolUrl(project, module, symbol),
+    });
+  });
+};
+
+const generateModuleDescriptors = (
+  descriptors: Array<SearchDescriptor>,
+  project: Project
+) => {
+  project.modules.forEach((module) => {
+    descriptors.push({
+      name: module.name,
+      type: "module",
+      url: generateModuleUrl(project, module),
+    });
+
+    if (module.functions?.length > 0) {
+      generateSymbolDescriptors(
+        descriptors,
+        project,
+        module,
+        module.functions,
+        "function"
+      );
+    }
+    if (module.variables?.length > 0) {
+      generateSymbolDescriptors(
+        descriptors,
+        project,
+        module,
+        module.variables,
+        "variable"
+      );
+    }
+    if (module.classes?.length > 0) {
+      generateSymbolDescriptors(
+        descriptors,
+        project,
+        module,
+        module.classes,
+        "class"
+      );
+    }
```

Comment:
These fields are guaranteed to exist
```suggestion
    generateSymbolDescriptors(
      descriptors,
      project,
      module,
      module.functions,
      "function"
    );
    generateSymbolDescriptors(
      descriptors,
      project,
      module,
      module.variables,
      "variable"
    );
    generateSymbolDescriptors(
      descriptors,
      project,
      module,
      module.classes,
      "class"
    );
```

Reply:
Cool! Fixed in 4af269e173a0d4133ca09daaad174b0e10f1d951